### PR TITLE
feat(cdn): store logs in buffer unit (#5353)

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Review by richardlt — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

File: sdk/error.go [hunk 1]
```diff
@@ -198,6 +198,7 @@ var (
 	ErrUnsupportedMediaType                          = Error{ID: 188, Status: http.StatusUnsupportedMediaType}
 	ErrNothingToPush                                 = Error{ID: 189, Status: http.StatusBadRequest}
 	ErrWorkerErrorCommand                            = Error{ID: 190, Status: http.StatusBadRequest}
+	ErrConflictData                                  = Error{ID: 191, Status: http.StatusConflict}
```
Comment (yesnault) [inline sdk/error.go:201]:
```suggestion
	ErrConflictData                                  = Error{ID: 192, Status: http.StatusConflict}
```

**Review by yesnault — CHANGES_REQUESTED** [review]

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14176.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘  
  * Standalone tests 1 -   
  * Standalone tests 2 -   
  * Standalone tests 3 -

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14217.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14222.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/cache/cache.go [hunk 2]
```diff
@@ -68,7 +68,9 @@ type LockStore interface {
 
 type ScoredSetStore interface {
 	ScoredSetAdd(ctx context.Context, key string, value interface{}, score float64) error
+	ScoredAppend(ctx context.Context, key string, value interface{}) error
```
Comment (fsamin) [inline engine/cache/cache.go:71]: ScoredSetAppend

File: engine/cache/cache.go [hunk 3]
```diff
@@ -68,7 +68,9 @@ type LockStore interface {
...
 	ScoredSetScan(ctx context.Context, key string, from, to float64, dest interface{}) error
+	SetCard(key string) (int, error)
```
Comment (fsamin) [inline engine/cache/cache.go:73]: ScoredSetCard

File: engine/cache/redis.go [hunk 4]
```diff
@@ -500,6 +500,22 @@ func (s *RedisStore) Unlock(key string) error {
 	return s.Delete(key)
 }
 
+func (s *RedisStore) ScoredAppend(ctx context.Context, key string, value interface{}) error {
```
Comment (fsamin) [inline engine/cache/redis.go:503]: ScoredSetAppend

**Review by richardlt — APPROVED** [review]

File: engine/cdn/storage/redis/redis.go [hunk 5]
```diff
@@ -40,13 +40,30 @@ func (s *Redis) ItemExists(i index.Item) (bool, error) {
 }
 
 func (s *Redis) Add(i index.Item, index uint, value string) error {
+	value = fmt.Sprintf("%d#%s", index, value)
```
Comment (fsamin) [inline engine/cdn/storage/redis/redis.go:43]:
```suggestion
	value = strconv.Itoa(int(index)) + "#" + value
```

File: engine/cdn/index/types.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
+package index
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"github.com/ovh/cds/engine/gorpmapper"
```
Comment (yesnault) [inline engine/cdn/index/types.go:7]: order?

File: engine/cdn/storage/redis/redis.go [hunk 7]
```diff
@@ -76,11 +93,19 @@ func (r *reader) Read(p []byte) (n int, err error) {
 	if err != nil {
 		return 0, err
 	}
+
 	if len(lines) > 0 {
+		cleanedLines := make([]string, 0, len(lines))
+		for _, s := range lines {
+			linesSplitted := strings.SplitAfterN(s, "#", 2)
```
Comment (fsamin) [inline engine/cdn/storage/redis/redis.go:100]: this split should be only in `Get`

File: engine/cdn/storage/redis/redis.go [hunk 8]
```diff
@@ -40,13 +40,30 @@ func (s *Redis) ItemExists(i index.Item) (bool, error) {
...
+
 func (s *Redis) Get(i index.Item, from, to uint) ([]string, error) {
 	var res = make([]string, to-from+1)
-	err := s.store.ScoredSetScan(context.Background(), i.ID, float64(from), float64(to), &res)
-	return res, err
+	if err := s.store.ScoredSetScan(context.Background(), i.ID, float64(from), float64(to), &res); err != nil {
+		return res, err
+	}
+	for i := range res {
+		linesSplitted := strings.SplitAfterN(res[i], "#", 2)
```
Comment (fsamin) [inline engine/cdn/storage/redis/redis.go:61]: I wonder is this is more efficient:
res[i] = strings.TrimLeftFunc(res[i], func(r rune) bool {
     return unicode.IsDigit(r) || unicode.IsNumber(r) || r == rune('#')
})

**Review by yesnault — CHANGES_REQUESTED** [review]

**Review by fsamin — CHANGES_REQUESTED** [review]

File: engine/cache/cache.go (hunk 3 above)
Comment (sguiheux) [reply engine/cache/cache.go:73]: SetCard already exist, don't want to duplicate implementation

File: engine/cdn/storage/redis/redis.go (hunk 8 above)
Comment (sguiheux) [reply engine/cdn/storage/redis/redis.go:61]: Will use:  
res[i] = strings.TrimFunc(res[i], unicode.IsNumber)
res[i] = strings.TrimPrefix(res[i], "#")

**Review by fsamin — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

**Pull request merged**